### PR TITLE
[SYCLomatic] Avoid migration of non-CUDA type "half"

### PR DESCRIPTION
**Pull request by the-slow-one**
Problem: `half` defined in a location outside of source under investigation and outside of CUDA header is migrated to `sycl::half`.

This PR prevents such migration.

### CODE REVIEW

File: clang/lib/DPCT/RulesLang/RulesLang.cpp [hunk 1]
```diff
@@ -1392,7 +1392,7 @@ void TypeInDeclRule::runRule(const MatchFinder::MatchResult &Result) {
           return;
 
         // When a CUDA type is redefined in the files under analysis we
-        // want to migrate them.
+        // don't want to migrate them.
```
Comment (zhiweij1) [inline clang/lib/DPCT/RulesLang/RulesLang.cpp]: The original comment is correct. We need to migrate redeclaration of CUDA type.

File: clang/lib/DPCT/RulesLang/RulesLang.cpp [hunk 2]
```diff
@@ -1753,11 +1758,15 @@ void VectorTypeNamespaceRule::runRule(const MatchFinder::MatchResult &Result) {
     if (TL->getBeginLoc().isInvalid())
       return;
 
-    // To skip user-defined type.
+    // To skip user-defined type (found in in-root and from third party includes outside of in-root.
```
Comment (zhiweij1) [inline clang/lib/DPCT/RulesLang/RulesLang.cpp]:
```suggestion
    // To skip user-defined type (found in in-root and from third party includes outside of in-root)
```

File: clang/lib/DPCT/RulesLang/RulesLang.cpp [hunk 3]
```diff
@@ -1540,6 +1541,10 @@ void TypeInDeclRule::runRule(const MatchFinder::MatchResult &Result) {
           return;
         }
       } else if (NTL.getTypeLocClass() == clang::TypeLoc::Record) {
+        if (!DpctGlobalInfo::isInCudaPath(NTL.getBeginLoc())) {
+            return;
```
Comment (zhiweij1) [inline clang/lib/DPCT/RulesLang/RulesLang.cpp:1545]:
```suggestion
          return;
```

File: clang/lib/DPCT/RulesLang/RulesLang.cpp [hunk 4]
```diff
@@ -1522,7 +1522,7 @@ void TypeInDeclRule::runRule(const MatchFinder::MatchResult &Result) {
       auto ETC = TL->getAs<ElaboratedTypeLoc>();
       auto NTL = ETC.getNamedTypeLoc();
 
-      if (const auto *RD = NTL.getType().getCanonicalType()->getAsRecordDecl())
+      if (const auto *RD = NTL.getType().getCanonicalType()->getAsRecordDecl()) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/RulesLang/RulesLang.cpp]:
```suggestion
      if (const auto *RD = NTL.getType().getCanonicalType()->getAsRecordDecl())
```

File: clang/lib/DPCT/RulesLang/RulesLang.cpp [hunk 5]
```diff
@@ -1531,6 +1531,7 @@ void TypeInDeclRule::runRule(const MatchFinder::MatchResult &Result) {
             if (const auto *Init = VarD->getInit())
               if (const auto *VarInitExpr = dyn_cast<InitListExpr>(Init))
                 emplaceTransformation(new ReplaceStmt(VarInitExpr, "{}"));
+      }
```
Comment (zhiweij1) [inline clang/lib/DPCT/RulesLang/RulesLang.cpp]:
```suggestion
```

File: clang/lib/DPCT/RulesInclude/InclusionHeaders.cpp [hunk 6]
```diff
@@ -299,4 +299,4 @@ DpctInclusionHeadersMap::DpctInclusionHeadersMapInitializer
     DpctInclusionHeadersMap::Initializer;
 
 } // namespace dpct
-} // namespace clang
\ No newline at end of file
+} // namespace clang
```
Comment (zhiweij1) [inline clang/lib/DPCT/RulesInclude/InclusionHeaders.cpp]: Please do not touch unrelated file.

File: clang/lib/DPCT/RulesLang/RulesLang.cpp (hunk 1 above)
Comment (zhiweij1) [reply clang/lib/DPCT/RulesLang/RulesLang.cpp]: How about add
```
      } else if (!isTypeInCUDAScope(TypePtr)) {
        return;
      }
```
after line 1399?

File: clang/lib/DPCT/RulesLang/RulesLang.cpp [hunk 7]
```diff
@@ -1540,6 +1541,10 @@ void TypeInDeclRule::runRule(const MatchFinder::MatchResult &Result) {
           return;
         }
       } else if (NTL.getTypeLocClass() == clang::TypeLoc::Record) {
+        if (!DpctGlobalInfo::isInCudaPath(NTL.getBeginLoc())) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/RulesLang/RulesLang.cpp]: How about changing at line 1402? (See above comment)

**Review by zhiweij1 — CHANGES_REQUESTED** [review]
Please add a lit test.

File: clang/lib/DPCT/RulesLang/RulesLang.cpp (hunk 1 above)
Comment (the-slow-one) [reply clang/lib/DPCT/RulesLang/RulesLang.cpp]: The "half" is outside the analysis scope. A test is added now which will make this clearer. At L1378 there is a check `isTypeInAnalysisScope`.

But I do get your point! Ideally, we should perform this check for all types declared in CUDA header and exit this method early. I will try this and see what break.

Comment (zhiweij1) [reply clang/lib/DPCT/RulesLang/RulesLang.cpp]: I mean the `else if (!isTypeInCUDAScope(TypePtr))` is added after the block of `if (isTypeInAnalysisScope(TypePtr)) { ... }`.

Comment (the-slow-one) [reply clang/lib/DPCT/RulesLang/RulesLang.cpp]: Done

**Review by zhiweij1 — APPROVED** [review]

**Review by TejaX-Alaghari — APPROVED** [review]
LGTM
